chamber/models/fields.py

```diff
@@ -138,23 +138,34 @@ def __init__(self, *args, **kwargs):
             kwargs['choices'] = self.enum.choices
         super(SubchoicesPositiveIntegerField, self).__init__(*args, **kwargs)
 
+    def _get_subvalue(self, model_instance):
+        return getattr(model_instance, self.subchoices_field_name)
+
     def clean(self, value, model_instance):
-        supvalue = getattr(model_instance, self.subchoices_field_name)
-        if self.enum and supvalue not in self.enum.categories:
+        if self.enum and self._get_subvalue(model_instance) not in self.enum.categories:
             return None
         else:
             return super(SubchoicesPositiveIntegerField, self).clean(value, model_instance)
 
+    def _raise_error_if_value_suppose_to_be_empty(self, value, subvalue):
+        if self.enum and subvalue not in self.enum.categories and value is not None:
+            raise ValidationError(ugettext('Value must be empty'))
+
+    def _raise_error_if_value_is_not_allowed(self, value, subvalue, model_instance):
+        allowed_values = self.enum.get_allowed_states(getattr(model_instance, self.subchoices_field_name))
+        if subvalue in self.enum.categories and value not in allowed_values:
+            raise ValidationError(ugettext('Allowed choices are {}.').format(
+                ', '.join(('{} ({})'.format(*(self.enum.get_label(val), val)) for val in allowed_values))
+            ))
+
     def validate(self, value, model_instance):
-        if self.enum:
-            supvalue = getattr(model_instance, self.subchoices_field_name)
-            allowed_values = self.enum.get_allowed_states(getattr(model_instance, self.subchoices_field_name))
-            if self.enum and supvalue not in self.enum.categories and value is not None:
-                raise ValidationError(ugettext('Value must be empty'))
-            elif supvalue in self.enum.categories and value not in allowed_values:
-                raise ValidationError(ugettext('Allowed choices are {}.').format(
-                    ', '.join(('{} ({})'.format(*(self.enum.get_label(val), val)) for val in allowed_values))
-                ))
+        if not self.enum:
+            return
+
+        subvalue = self._get_subvalue(model_instance)
```

Comment:
Možná bych tu metodu 2x zavolal než dělat proměnnou, ale je to asi otázka vkusu...